### PR TITLE
Changes for supporting WPE Browser

### PR DESCRIPTION
**Pull request by psivasubramanian**
Changes for cdmiservice to work with WPE Browser after addressing the review comments from https://github.com/linaro-home/open-content-decryption-module-cdmi/pull/7

### CODE REVIEW

**Review by andrey-konovalov — COMMENTED** [review]
I am still not convinced that changes to rpc/opencdm_xdr.h are needed.

Just adding the #defines to rpc/opencdm_xdr.h doesn't seem to make sense, as these #define's aren't used - both opencdm_xdr_clnt.c and opencdm_xdr_srv.c ignore them. Compare it to e.g. how RPC_OPEN_CDM_MEDIAKEYSESSION_RELEASE is handled:
----- 8<-----
$ grep RPC_OPEN_CDM_MEDIAKEYSESSION_RELEASE -r . -n
./rpc/opencdm_xdr.h:139:#define RPC_OPEN_CDM_MEDIAKEYSESSION_RELEASE 6
./rpc/opencdm_xdr.h:167:#define RPC_OPEN_CDM_MEDIAKEYSESSION_RELEASE 6
./rpc/opencdm_xdr_clnt.c:93:	if (clnt_call (clnt, RPC_OPEN_CDM_MEDIAKEYSESSION_RELEASE,
./rpc/opencdm_xdr.x:88:    rpc_response_generic RPC_OPEN_CDM_MEDIAKEYSESSION_RELEASE(rpc_request_session_release) = 6;
./rpc/opencdm_xdr_svc.c:70:	case RPC_OPEN_CDM_MEDIAKEYSESSION_RELEASE:
$
----- 8<-----

Also I took a look at where the files under rpc/ came from. The primary sources there are opencdm_callback.x and opencdm_xdr.x - these files must be modified if changes to rpc are needed. The rest of the files (originally all under rpc/gen/) are generated by rpcgen, and may not be hand edited. Later, it was commit beb60b2 "Added RPC headers and implementations" ("We don't have to generate the RPC headers every time") which under /rpc added *the copies* of the generated files.

**Comment by andrey-konovalov** [comment]
Otherwise the new patch set looks much better than the 1st version. The other three commits look OK to me.

**Comment by andrey-konovalov** [comment]
As a side note, the #defines which Siva is adding, look like came from following the (more generic definitions) at https://w3c.github.io/encrypted-media/#mediakeysession-interface (see the WebIDL insert):
```
    Promise<void>    generateRequest(DOMString initDataType,
                                     BufferSource initData);
    Promise<boolean> load(DOMString sessionId);
    Promise<void>    update(BufferSource response);
    Promise<void>    close();
    Promise<void>    remove();
```
and at https://w3c.github.io/encrypted-media/#mediakeys-interface
```
interface MediaKeys {
    MediaKeySession  createSession(optional MediaKeySessionType sessionType = "temporary");
    Promise<boolean> setServerCertificate(BufferSource serverCertificate);
};
```
While the current implementation is more following the https://w3c.github.io/encrypted-media/#clear-key (including "The setServerCertificate() method: Not supported." statement). And at my first look without supporting persistent sessions support close() and remove() doesn't differ, and could be named "release()" as well.

**Comment by andrey-konovalov** [comment]
(The side note continues)
I've also noticed that the current EME recommendation (both the draft, and the last released version) in the Abstract are now referencing HTML 5.1 (vs 5.0). I have no idea of the implications though

**Comment by psivasubramanian** [comment]
>Also I took a look at where the files under rpc/ came from. The primary sources there are opencdm_callback.x and opencdm_xdr.x - these files must be modified if changes to rpc are needed. The rest of the files (originally all under rpc/gen/) are generated by rpcgen, and may not be hand edited.

Ok, we actually made this RPC changes in opencdm_callback.x & opencdm_xdr.x as well previously. Since it's not reflecting, we reverted it.

> I've also noticed that the current EME recommendation (both the draft, and the last released version) in the Abstract are now referencing HTML 5.1 (vs 5.0). I have no idea of the implications though

Yes, we may need to implement these APIs when we move CDM to EMEv1/HTML 5.1 based.

File: cdmi-stub/mediakeysession.cpp [hunk 1]
```diff
@@ -154,10 +154,19 @@ void CMediaKeySession::Update(
...
                                    f_cbKeyMessageResponse);
-  //Session type is set to "0". We keep the function signature to
-  //match Chromium's ExtractKeysFromJWKSet(...) function
-  media::ExtractKeysFromJWKSet(key_string, &g_keys, 0);
-
+    
+  if (key_string.compare(0, 1, "{"))
+  {
+        CDMI_DLOG() << "KEY not in JSON format" << endl;
+        std::string key_id("1");       // dummy keyid for adding key_string to gkeys
```
Comment (petegriffin) [inline cdmi-stub/mediakeysession.cpp:161]: Can you be more verbose about why this change is required?

Specifically things like:
1) What format is the key being received in?
2) What is sending the key in this format (EME test app, WPE)?

File: cdmi-stub/mediakeys.cpp [hunk 2]
```diff
@@ -35,7 +35,12 @@ CMediaKeys::~CMediaKeys(void) {}
...
     const char *f_pwszMimeType,
     const char *f_pwszKeySystem) const {
-  bool isSupported = true;
+  bool isSupported = false;
+  if (f_pwszKeySystem)
+      if (strcmp(f_pwszKeySystem, "org.w3.clearkey") == 0 || strcmp(f_pwszKeySystem, "org.chromium.externalclearkey") == 0)
+          isSupported = true;
+
+  return isSupported;
 }
```
Comment (petegriffin) [inline cdmi-stub/mediakeys.cpp:45]: LGTM

File: rpc/opencdm_xdr.h [hunk 3]
```diff
@@ -143,6 +143,10 @@ extern  rpc_response_generic * rpc_open_cdm_mediakeysession_release_1_svc(rpc_re
 extern  rpc_response_generic * rpc_open_cdm_mediaengine_1(rpc_request_mediaengine_data *, CLIENT *);
 extern  rpc_response_generic * rpc_open_cdm_mediaengine_1_svc(rpc_request_mediaengine_data *, struct svc_req *);
 extern int open_cdm_1_freeresult (SVCXPRT *, xdrproc_t, caddr_t);
+//WPE
+#define RPC_OPEN_CDM_MEDIAKEYS_SET_SERVER_CERTIFICATE 8
+#define RPC_OPEN_CDM_MEDIAKEYSESSION_REMOVE 9
+#define RPC_OPEN_CDM_MEDIAKEYSESSION_CLOSE 10
```
Comment (petegriffin) [inline rpc/opencdm_xdr.h:149]: Are these new defines used by anything currently?

I guess this is the new RPC interface changes to match what was added in OpenCDM for the new EME APIs?

It looks like we also need a minimal implementation for these, or at least a stub to print "not implemented" or similar.

With this commit it would also be good to get a more verbose commit message, why is it required?

Comment (psivasubramanian) [reply rpc/opencdm_xdr.h:149]: >>Are these new defines used by anything currently?
No, it's not used by anything.

>>I guess this is the new RPC interface changes to match what was added in OpenCDM for the >>new EME APIs?
Yes.

>>It looks like we also need a minimal implementation for these, or at least a stub to print "not >>implemented" or similar.
>>With this commit it would also be good to get a more verbose commit message, why is it >>required?
default case handles that with some message.

File: cdmi-stub/mediakeysession.cpp (hunk 1 above)
Comment (psivasubramanian) [reply cdmi-stub/mediakeysession.cpp:161]: WPE/YouTube EME is sending the key in non-json format whereas Chromium EME app sends in json format, that why this is required.
